229: Flight targets form refactor

rpeterman-gp opened the pull request:
Closes #229 

This PR adds a menu to the __Add Target__ button of the flight targets form component. The menu provide the types of targets to add. Once clicked, a new empty target locked to that type will be added to the targets list.

This also refactors form value change handling to utilize the FormBuilder events/handlers, paying close attention not to trigger more than one update cycle per change. This avoids having to create and wire up emitters for ad/remove target actions.

rpeterman-gp committed:
feat(campaign): add actions to add/remove target

- wire up event emitters for adding/removing flight target
- add selectors to select target types from new API model
- add selector to select targets mapped by type
- add menu to targets form to add target by type

rpeterman-gp committed:
fix(flight-targets-form): refactor change handling

- add selectors for inventory target
- pass target options map through flight form components
- rework target options to use map object
- update tests
- styling changes to account for lose of target type input
- TODO: remove extra debugging code (lots of console.logs)

sandikbarr commented on src/app/campaign/flight/flight-form.component.scss:
```diff
@@ -1,6 +1,14 @@
 @import '~src/sass/variables';
 @import '~src/sass/colors';
 
+:host {
```
nit: with css encapsulation by default in Angular, the style is always scoped to the host. In general I'm not a fan of Sass scoping when it's not necessary. It locks us into Sass, and when we do use it, I often think it should be a smaller component instead. These large style sheets can make refactoring more difficult. /rant

rpeterman-gp replied:
Yeah. I always forget Angular does this for us. Will remove.

sandikbarr commented on src/app/campaign/flight/flight-targets-form.component.ts:
```diff
@@ -89,64 +165,81 @@ export class FlightTargetsFormComponent implements ControlValueAccessor {
-  private setCodeOptions(targets: FlightTarget[]) {
-    this.codeOptions = (targets || []).map(target => {
-      return this.filteredTargetOptions(target.type, target.code, targets);
+    newIndex = this.targets.length;
```
is newIndex used?

rpeterman-gp replied:
Yep. Removed.

rpeterman-gp committed:
refactor: remove add/remove target actions

- clean up debugging code
- clean up commented code
- update flight targets form tests
- add getter/setter for targetOptionsMap
- move check for empty target codes to flight form change stream

rpeterman-gp committed:
feat(flight-targets-form): add ngOnDestroy

- unsubscribe from targetsForm changes when compoent is destroyed

rpeterman-gp committed:
Merge branch 'master' into feat/132-episode-search-component

rpeterman-gp committed:
fix(flight-targets-form): episode label format

- create prepended date string using utc.

sandikbarr commented on src/app/campaign/flight/flight-targets-form.component.ts:
```diff
@@ -41,44 +38,87 @@ import moment from 'moment';
-    });
+    // TODO: This check is here to prevent an error seen in tests. Determine if
+    // there is something in the spec that could be done differently to not
+    // require this check.
```
weird, wonder what the deal is

sandikbarr commented on src/app/campaign/flight/flight-zones-form.component.ts:
```diff
@@ -57,7 +57,7 @@ export const validateMp3 = (value: string): { [key: string]: any } | null => {
       </div>
     </fieldset>
     <div *ngIf="zones?.controls?.length < zoneOptions?.length">
-      <a mat-button routerLink="." (click)="onAddZone()"> <mat-icon>add</mat-icon> {{ addCreativeLabel }} </a>
```
good call, I had just copied the previous zones markup, but buttons should be buttons.

sandikbarr commented on src/sass/styles.scss:
```diff
@@ -14,6 +14,10 @@ body {
   font-family: Roboto, 'Helvetica Neue', sans-serif;
 }
 
+button.mat-button {
```
👍

sandikbarr reviewed (approved):
awesome, gonna go ahead and get this merged 🎉

sandikbarr commented:
or actually is this still in draft/not ready? anyway lookin good
